File: topics/HYPERLOGLOG/hyperloglog.py
import hashlib
import math

class HyperLogLog:
    """
    A probabilistic data structure used to estimate the cardinality (number of unique elements) of a set.
    """
    def __init__(self, b):
        """
        Initialize HLL.
        b : int : the number of bits used to determine the bucket (register) index.
                  Total buckets m = 2^b.
                  Standard error is roughly 1.04 / sqrt(m).
        """
        self.b = b
        self.m = 1 << b # 2^b buckets
        # Initialize all buckets to 0
        self.registers = [0] * self.m

    def _get_hash(self, item):
        """Returns a 32-bit hash of the item as an integer."""
        digest = hashlib.sha256(str(item).encode('utf-8')).hexdigest()
        return int(digest[:8], 16) # use first 8 hex characters (32 bits)

    def _get_leading_zeros(self, val, max_bits):
        """Counts the number of leading zeros in the binary representation."""
        binary = bin(val)[2:].zfill(max_bits)
        count = 0
        for bit in binary:
            if bit == '0':
                count += 1
            else:
                break
        return count + 1 # Include the first '1' as part of the run length

    def add(self, item):
        """
        Adds an item to the HyperLogLog structure.
        """
        # 1. Hash the item to 32 bits
        h = self._get_hash(item)

        # 2. Use the first 'b' bits as the bucket index
        # Shift right by (32 - b) to get the top 'b' bits
        bucket_index = h >> (32 - self.b)

        # 3. Use the remaining bits to calculate leading zeros
        # Create a mask for the remaining (32 - b) bits
        mask = (1 << (32 - self.b)) - 1
        remaining_bits = h & mask

        # 4. Count leading zeros
        leading_zeros = self._get_leading_zeros(remaining_bits, 32 - self.b)

        # 5. Update the register if the new leading zeros count is higher
        self.registers[bucket_index] = max(self.registers[bucket_index], leading_zeros)

    def count(self):
        """
        Calculates the estimated cardinality using the Harmonic Mean of the registers.
        """
        # Calculate Harmonic Mean sum
        z = sum([2.0 ** -val for val in self.registers])

        # Calculate alpha_m correction factor
        if self.m == 16:
            alpha = 0.673
        elif self.m == 32:
            alpha = 0.697
        elif self.m == 64:
            alpha = 0.709
        else:
            alpha = 0.7213 / (1.0 + 1.079 / self.m)

        # Raw estimate E = alpha_m * m^2 * (1 / Z)
        estimate = alpha * (self.m ** 2) / z

        # Apply small range correction if needed
        if estimate <= 2.5 * self.m:
            zeros = self.registers.count(0)
            if zeros > 0:
                # Linear counting
                estimate = self.m * math.log(self.m / zeros)

        return int(estimate)

if __name__ == '__main__':
    # Sample Input: Using b=10 (1024 buckets). Expected error ~ 1.04/sqrt(1024) = 3.25%
    hll = HyperLogLog(b=10)

    print("Inserting 10,000 unique items...")
    # Insert 10,000 unique integers
    for i in range(10000):
        hll.add(f"user_{i}")

    # Also insert 5,000 duplicates
    for i in range(5000):
        hll.add(f"user_{i}")

    # Sample Output
    estimated_count = hll.count()
    actual_count = 10000
    print(f"Actual Unique Count: {actual_count}")
    print(f"HLL Estimated Count: {estimated_count}")
    print(f"Error Rate: {abs(actual_count - estimated_count) / actual_count * 100:.2f}%")

File: topics/HYPERLOGLOG/hyperloglog.rs
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// HyperLogLog probabilistic cardinality estimator
pub struct HyperLogLog {
    /// Number of bits used for indexing the buckets
    b: u32,
    /// Number of buckets (m = 2^b)
    m: usize,
    /// The buckets storing the max leading zeros
    registers: Vec<u8>,
}

impl HyperLogLog {
    /// Initialize a new HLL with b bits for the bucket index
    pub fn new(b: u32) -> Self {
        assert!(b >= 4 && b <= 16, "b must be between 4 and 16");
        let m = 1_usize << b;
        HyperLogLog {
            b,
            m,
            registers: vec![0; m],
        }
    }

    /// Hash the item into a 64-bit integer
    fn get_hash<T: Hash>(&self, item: &T) -> u64 {
        let mut hasher = DefaultHasher::new();
        item.hash(&mut hasher);
        hasher.finish()
    }

    /// Add an item to the HLL
    pub fn add<T: Hash>(&mut self, item: &T) {
        let hash = self.get_hash(item);

        // Use the highest 'b' bits to find the bucket index
        let bucket_index = (hash >> (64 - self.b)) as usize;

        // Use the remaining lower bits to count leading zeros.
        // We isolate them by shifting left by 'b' bits, then counting leading zeros.
        // Note: we add 1 because we want the position of the first 1-bit.
        let remaining_bits = hash << self.b;
        let leading_zeros = remaining_bits.leading_zeros() as u8 + 1;

        // Keep the maximum value seen in this bucket
        if leading_zeros > self.registers[bucket_index] {
            self.registers[bucket_index] = leading_zeros;
        }
    }

    /// Estimate the number of unique elements
    pub fn count(&self) -> usize {
        // Calculate the harmonic mean
        let mut z = 0.0_f64;
        for &val in &self.registers {
            z += 2.0_f64.powi(-(val as i32));
        }

        // Apply the correction factor (alpha_m)
        let alpha = match self.m {
            16 => 0.673,
            32 => 0.697,
            64 => 0.709,
            _ => 0.7213 / (1.0 + 1.079 / (self.m as f64)),
        };

        // Raw estimate
        let mut estimate = alpha * (self.m.pow(2) as f64) / z;

        // Small range correction (Linear Counting)
        if estimate <= 2.5 * (self.m as f64) {
            let zeros = self.registers.iter().filter(|&&v| v == 0).count();
            if zeros > 0 {
                estimate = (self.m as f64) * ((self.m as f64) / (zeros as f64)).ln();
            }
        }

        estimate.round() as usize
    }
}

fn main() {
    // b=10 means 1024 buckets
    let mut hll = HyperLogLog::new(10);

    println!("Inserting 10,000 unique items...");
    // Insert 10,000 unique integers
    for i in 0..10000 {
        hll.add(&format!("user_{}", i));
    }

    // Insert 5,000 duplicate integers
    for i in 0..5000 {
        hll.add(&format!("user_{}", i));
    }

    let estimated_count = hll.count();
    let actual_count = 10000;

    println!("Actual Unique Count: {}", actual_count);
    println!("HLL Estimated Count: {}", estimated_count);

    let error_rate = ((actual_count as f64 - estimated_count as f64).abs() / actual_count as f64) * 100.0;
    println!("Error Rate: {:.2}%", error_rate);
}
